irods/manager/util/generic_api_endpoint.py

```diff
@@ -0,0 +1,12 @@
+from irods.api_number import api_number
+from irods.message import iRODSMessage, JSON_Message
+
+def touch_impl(session, path, **options):
```

Comment:
Good idea. Keep them conventions coming.

Reply:
Thinking about this more, user code shouldn't import the `util` module at all. Are there conventions for expressing that?

Reply:
@korydraughn - prefixing with underscore, ie renaming to `_touch_impl` as you said., is probably what you want That's a clear enough indication to users this isn't for them to call.  Also (and in support of that conclusion), a statement `from <module> import *` won't import symbols starting with underscores

Reply:
(One can still do `from X.Y.Z import _my_prefixed_thing`) or just  `import X.Y.Z` then then do something with X.Y.Z._my_prefixed_thing

Reply:
That means all functions within the module need a leading underscore. Is it common to see modules containing only functions with leading underscores?

Reply:
I'm not sure. You could also go the Compose route and just mark the module as "not intended to be an API" :)

In all seriousness, maybe there's no benefit to trying to hide the function or discourage its use other than leaving a comment to that effect in the docstring or the documentation.

Reply:
Yeah. It's not clear which route to take. I may just add a comment to the module.

> You could also go the Compose route and just mark the module as "not intended to be an API" :)

Is that something that causes python to issue a warning somewhere?
Does Python have docstrings for entire source files?

Reply:
No, they just left a comment in the README (which has now been removed since they abandoned the Python implementation) buried in some other stuff. I'm not bitter at all. :p I defer to @d-w-moore. Seems like this area is Wild West.

Reply:
@korydraughn  - You could set __all__ in that module to an empty list...
```
__all__ = [ ]
```
 In general , if you want to dictate how symbols are mass-imported from a module , the solution is probably to enumerate them in the __all__ array. that's the only other rule / convention I know of to govern this sort of thing....

For example,
```
#mymodule
__all__ = ["x"]
x=1
y=2
```
Then: from mymodule import * only imports `x`, not `y`
Hope that helps.  beyond that, I am not left with much guidance I can give.

Reply:
As usual, you can't perform perfect data hiding in Python. You can't prevent for example
```
from your_module import _symbol_you_would_prefer_to_be_outside_user_access
```

Reply:
Well, the worst part is that we do intend for this to be imported... by other modules in the library. So, I'm thinking we put a leading underscore and a note in the docstring. Thoughts?

Reply:
> Is that something that causes python to issue a warning somewhere? 

I don't think you can issue a warning specifically to user code for importing something, but maybe if I had a few hours to research or think it out, I could hack something together.

> Does Python have docstrings for entire source files?
Yes.  The following module has a docstring that shows up in `pydoc3 mymodule_with_doc`
# mymodule_with_doc.py
```
"""This is the README docstring for this module.
"""
def useless_function( ): pass
```

Reply:
> Well, the worst part is that we do intend for this to be imported... by other modules in the library. So, I'm thinking we put a leading underscore and a note in the docstring. Thoughts?

Seems fine.  As stated, nothing prevents other modules from importing symbols in your module.  You just have to do it explicitly.  i.e. by mentioning the symbol, not using `*`

Reply:
Got it. Going to move forward with a comment in the source files.

Reply:
@alanking in other words doing `__all__ = [ ]` does not prevent individual import of symbols by name, nor does it prevent access to the module namespace.

Reply:
I'm just now seeing there were additional comments. My comment about moving forward with a source file level docstring didn't take the rest of the conversation into account.

I'm going to add leading underscores.

Reply:
I can also rename the module to use a leading underscore. That seems to be a known convention.

Thoughts?

Reply:
Done